### ISSUE TITLE
What is the exact behaviour of undef<T>()?

### ISSUE THREAD

**Issue by Poutsh** [halide/Halide#3011]
Hi, I was wondering what is the exact behaviour of using undef<T>() on a pure definition of a function.

In my case, my input buffers have a huge size and I use a variable to decide how much of the buffer I want to compute. So I initialize the pure definition with undef<T>() and use a RDom with my size variable.

My code is like that : 
```c++
RDom(0, size);
Func f;
f(x) = undef<float>();
f(r) = some value;
```

Does undef acts like nothing is done in the x loop so that it only computes f on my reduction domain? If not, is there a way to do it in Halide?
Thank you in advance.

**Comment by zvookin** [comment]
tl;dr Yes.

undef elides the initialization. I think there is checking to make sure the
next stage does not read the undef'ed area, but it may be left as undefined
behavior. There is nothing to check that the full computation is fulfilled.

Example program:

#include "Halide.h"

using namespace Halide;

int main(int argc, char **argv) {
    Var x;
    Func f;
    RDom r(5, 10);

    f(x) = undef<float>();
    f(r) = 42.0f;

    f.realize(100);
    Target t("host-no_asserts-no_bounds_query");
    f.compile_to_lowered_stmt("/tmp/undef_partial_rdom.stmt", {},
StmtOutputFormat::Text, t);
}


Stmt output:

module name=f,
target=x86-64-osx-avx-avx2-f16c-fma-no_asserts-no_bounds_query-sse41
external_plus_metadata func f (f) {
assert((reinterpret(uint64, f.buffer) != (uint64)0),
halide_error_buffer_argument_is_null("f"))
let f = _halide_buffer_get_host(f.buffer)
let f.min.0 = _halide_buffer_get_min(f.buffer, 0)
produce f {
  for (f.s1.r$x, 5, 10) {
    f[(f.s1.r$x - f.min.0)] = 42.000000f
  }
}
}

-Z-
